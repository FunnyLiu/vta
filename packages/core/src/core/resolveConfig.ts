import { Config } from "@vta/config";
import { appRunSync } from "./index";

export default function resolveConfig<T = Config>(key: string, cwd = process.cwd()): T {
  let resolve;
  appRunSync({ cwd, dontRun: true }, (err, appResolveConfig) => {
    resolve = appResolveConfig;
  });
  if (typeof resolve === "function") {
    return resolve(key);
  }
  return undefined;
}
