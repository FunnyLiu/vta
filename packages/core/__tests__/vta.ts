import path from "path";
import fs from "fs";
import appRun, { appRunSync } from "../src/core";
import resolveConfig from "../src/core/resolveConfig";

/* eslint-disable */
describe("vta-engine", () => {
  it("project-1", () => {
    process.chdir(path.resolve(__dirname, "data/project-1"));
    appRun().then(({ error: err }) => {
      expect(err).toBe(undefined);
      expect(
        JSON.stringify(JSON.parse(process.env.VTA_CONFIG_RECORD_PLUGIN_STORE), null, 2),
      ).toMatchSnapshot("config_record_plugin");
      expect(process.env.VTA_ADDITIONAL_PLUGIN_GUID).toBe("31131534-5e68-4b8d-96fa-757fbc2cc9b1");
      expect(process.env.VTA_ADDITIONAL_PLUGIN_RECORD_GUID).toBe(
        "905365f0-e49b-49bc-afe7-7550e46297bb",
      );
    });
  });

  it("project-2-invalid-vta-config", () =>
    appRunSync({ silent: true, cwd: path.resolve(__dirname, "data/project-2") }, err => {
      expect(!!err).toBe(true);
      expect(err.message.indexOf("cannot load vta config file") >= 0).toBe(true);
    }));

  it("project-3-invalid-plugin", () =>
    appRun({ silent: true, cwd: path.resolve(__dirname, "data/project-3") }).then(
      ({ error: err }) => {
        expect(!!err).toBe(true);
        expect(err.message.indexOf("cannot load plugin") >= 0).toBe(true);
      },
    ));

  it("project-4-empty-plugins", () =>
    appRun({ silent: true, cwd: path.resolve(__dirname, "data/project-4") }).then(
      ({ error: err }) => {
        expect(err).toBe(undefined);
      },
    ));

  it("project-5-plugin-exception", () =>
    appRun({ silent: true, cwd: path.resolve(__dirname, "data/project-5") }).then(
      ({ error: err }) => {
        expect(!!err).toBe(true);
        expect(err.message.indexOf("Plugin Promise Exception") >= 0).toBe(true);
      },
    ));

  it("project-6-restart", () => {
    process.env.NODE_ENV = "development";
    setTimeout(() => {
      fs.writeFileSync(
        path.resolve(__dirname, "data/project-6/config/01.temp.js"),
        "module.exports = {};",
      );
    }, 1000);
    return appRun({ silent: true, cwd: path.resolve(__dirname, "data/project-6") }).then(
      ({ error: err }) => {
        expect(err).toBe(undefined);
        expect(
          JSON.stringify(JSON.parse(process.env.VTA_CORE_PROJECT_6_STORE), null, 2),
        ).toMatchSnapshot("restart");
        const versions = JSON.parse(process.env.VTA_CORE_PROJECT_6_STORE_VERSIONS);
        expect(versions.length).toBe(2);
        expect(versions[0].indexOf("20191212-1")).toBe(0);
        expect(versions[1].indexOf("20191212-1")).toBe(0);
      },
    );
  });
});

describe("resolve-config", () => {
  it("project-1", () => {
    process.chdir(path.resolve(__dirname, "data/project-1"));
    const appConfig = resolveConfig("app");
    expect(appConfig.dirs.config).toBe("config-1");
    expect(appConfig.dirs.src).toBe("src");
    expect(appConfig.name).toBe("project-1");
    expect(appConfig.version).toBe("1.0.0");
    expect(appConfig.guid).toBe("905365f0-e49b-49bc-afe7-7550e46297bb");
  });

  it("project-3", () => {
    process.chdir(path.resolve(__dirname, "data/project-3"));
    const appConfig = resolveConfig("app");
    expect(appConfig).toBe(undefined);
  });
});
