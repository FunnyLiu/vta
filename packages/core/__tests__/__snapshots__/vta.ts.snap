// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vta-engine project-1: config_record_plugin 1`] = `
"{
  \\"processOrder\\": [
    \\"prepare\\",
    \\"apply\\",
    \\"hooks.ready\\",
    \\"hooks.config.app-base-start\\",
    \\"hooks.config.app-base-done\\",
    \\"hooks.config.app-user-start\\",
    \\"hooks.config.app-user-done\\",
    \\"hooks.config.app-done\\",
    \\"hooks.run\\",
    \\"hooks.done\\"
  ],
  \\"appConfig\\": {
    \\"baseValue\\": {
      \\"dirs\\": {
        \\"config\\": \\"config-1\\",
        \\"src\\": \\"src\\",
        \\"build\\": \\"build\\"
      },
      \\"name\\": \\"project-1-base\\",
      \\"version\\": \\"1.0.0\\"
    },
    \\"value\\": {
      \\"dirs\\": {
        \\"config\\": \\"config-1\\",
        \\"src\\": \\"src\\",
        \\"build\\": \\"build\\"
      },
      \\"name\\": \\"project-1\\",
      \\"version\\": \\"1.0.0\\",
      \\"guid\\": \\"905365f0-e49b-49bc-afe7-7550e46297bb\\"
    },
    \\"userGuid\\": \\"fa68896c-9eac-4aa0-8ac2-063a5aa3e3b1\\",
    \\"runGuid\\": \\"905365f0-e49b-49bc-afe7-7550e46297bb\\",
    \\"doneGuid\\": \\"905365f0-e49b-49bc-afe7-7550e46297bb\\"
  }
}"
`;

exports[`vta-engine project-6-restart: restart 1`] = `
"{
  \\"process\\": [
    \\"apply\\",
    \\"config-init\\",
    \\"ready\\",
    \\"config-app-done\\",
    \\"run\\",
    \\"restart\\",
    \\"apply\\",
    \\"config-init\\",
    \\"ready\\",
    \\"config-app-done\\",
    \\"run\\",
    \\"done\\"
  ],
  \\"versions\\": []
}"
`;
